shared/domains/chelonia/chelonia.js

```diff
@@ -108,6 +226,15 @@ export default (sbp('sbp/selectors/register', {
       return stack
     }
   },
+  'chelonia/withEnv': async function (contractID: string, env: Object, sbpInvocation: Array<*>) {
```

Comment:
You know, this `withEnv` thing seems a bit unclean to me.

For users of the framework to have to know when to use it and how to use it.. it's just too complicated.

Can you think of any way we can simplify things for them?

Imagine you're a user of Chelonia and you want to develop web apps with it. We want regular, ordinary web developers to be able to use it without too much fuss. How can we simplify things for them so that they don't have to worry about cryptography (and just use defaults that are defined elsewhere somehow), while allowing advanced users to still create their own protocols using their own ciphers should they want to?

Reply:
Current working plan to address this summarized in this comment: https://github.com/okTurtles/group-income/pull/1257#pullrequestreview-1172682503

Reply:
I think this was resolved but not I'm not sure, so leaving this open for now.

Reply:
> I think this was resolved but not I'm not sure, so leaving this open for now.

Many uses of `withEnv` have been moved elsewhere, yes.

> How can we simplify things for them so that they don't have to worry about cryptography (and just use defaults that are defined elsewhere somehow)

`withEnv` or an equivalent is needed to maintain a context in certain situations. SBP has a strong limitation in terms of contextual information, i.e., you can only have a shared global context that lives in `this`. Whenever some transient or contextual information is needed, this leaves three (four) possibilities:

* Something akin to `withEnv`
* Explicitly providing the context (as an argument to the selector)
* Storing this information in the shared context (similar to withEnv at this point) but manually
* (Something else I might be missing; such as a dedicated 'context' selector(?))

Any of these approaches have their merits, but I don't see how either will make 'providing a context' easier.

Then, there is the qualifier in the question: *worry __about cryptography__*

The protocol provides some cryptographic primitives but does not assign specific meanings to keys. This means that much of the task of 'writing a contract' requires worrying about cryptography at some level. We could add some abstractions on top of what we have for some common operations (kind of like what Redux Toolkit is to Redux), but ultimately I'd say that understanding how things work is a prerequisite to writing something close to a secure contract, because the specifics of what is being done matter.

> Imagine you're a user of Chelonia and you want to develop web apps with it. We want regular, ordinary web developers to be able to use it without too much fuss.

Besides the 'contract' part, you don't need to worry much about cryptography on the frontend side of things.

Reply:
Good points @corrideat 👍 

I think we'll stick with `withEnv` for now. I'm going to resolve this and open an other comment thread to track the deadlock issue we discussed on the call.